File: frontend/src/api/modules/video.ts
/**
 * 视频监控 API
 * Story 10-1: 摄像头元数据管理
 */
import request from '@/utils/request'
import type { PageParams, PageResponse } from './types'

// ==================== 类型定义 ====================

/** NVR 设备 */
export interface NVRItem {
  id: number
  name: string
  ip_address: string
  port: number
  username: string | null
  password_masked: string | null
  manufacturer: string | null
  model: string | null
  max_channels: number | null
  status: string
  description: string | null
  camera_count: number
  created_at: string | null
  updated_at: string | null
}

/** NVR 创建参数 */
export interface NVRCreateParams {
  name: string
  ip_address: string
  port?: number
  username?: string
  password?: string
  manufacturer?: string
  model?: string
  max_channels?: number
  description?: string
}

/** NVR 更新参数 */
export interface NVRUpdateParams {
  name?: string
  ip_address?: string
  port?: number
  username?: string
  password?: string
  manufacturer?: string
  model?: string
  max_channels?: number
  status?: string
  description?: string
}

/** 摄像头预置位 */
export interface CameraPreset {
  id: number
  camera_id: number
  preset_index: number
  name: string
  description: string | null
}

/** 预置位创建参数 */
export interface CameraPresetCreate {
  preset_index: number
  name: string
  description?: string
}

/** 摄像头 */
export interface CameraItem {
  id: number
  name: string
  code: string
  rtsp_url: string | null
  onvif_url: string | null
  hls_url: string | null
  nvr_id: number | null
  nvr_name: string | null
  channel_no: number | null
  area_code: string | null
  cabinet_id: number | null
  device_id: number | null
  location_description: string | null
  camera_type: string
  status: string
  is_enabled: boolean
  presets: CameraPreset[]
  created_at: string | null
  updated_at: string | null
}

/** 摄像头创建参数 */
export interface CameraCreateParams {
  name: string
  code: string
  rtsp_url?: string
  onvif_url?: string
  hls_url?: string
  nvr_id?: number
  channel_no?: number
  area_code?: string
  cabinet_id?: number
  device_id?: number
  location_description?: string
  camera_type?: string
  presets?: CameraPresetCreate[]
}

/** 摄像头更新参数 */
export interface CameraUpdateParams {
  name?: string
  code?: string
  rtsp_url?: string
  onvif_url?: string
  hls_url?: string
  nvr_id?: number | null
  channel_no?: number | null
  area_code?: string | null
  cabinet_id?: number | null
  device_id?: number | null
  location_description?: string
  camera_type?: string
  status?: string
  is_enabled?: boolean
  presets?: CameraPresetCreate[]
}

// ==================== NVR API ====================

/** 创建 NVR */
export function createNVR(data: NVRCreateParams): Promise<NVRItem> {
  return request.post('/v1/video/nvrs', data)
}

/** NVR 列表 */
export function getNVRList(params?: PageParams): Promise<PageResponse<NVRItem>> {
  return request.get('/v1/video/nvrs', { params })
}

/** NVR 详情 */
export function getNVR(id: number): Promise<NVRItem> {
  return request.get(`/v1/video/nvrs/${id}`)
}

/** 更新 NVR */
export function updateNVR(id: number, data: NVRUpdateParams): Promise<NVRItem> {
  return request.put(`/v1/video/nvrs/${id}`, data)
}

/** 删除 NVR */
export function deleteNVR(id: number): Promise<{ message: string }> {
  return request.delete(`/v1/video/nvrs/${id}`)
}

// ==================== Camera API ====================

/** 创建摄像头 */
export function createCamera(data: CameraCreateParams): Promise<CameraItem> {
  return request.post('/v1/video/cameras', data)
}

/** 摄像头列表 */
export function getCameraList(params?: PageParams & {
  nvr_id?: number
  area_code?: string
  status?: string
}): Promise<PageResponse<CameraItem>> {
  return request.get('/v1/video/cameras', { params })
}

/** 摄像头详情 */
export function getCamera(id: number): Promise<CameraItem> {
  return request.get(`/v1/video/cameras/${id}`)
}

/** 更新摄像头 */
export function updateCamera(id: number, data: CameraUpdateParams): Promise<CameraItem> {
  return request.put(`/v1/video/cameras/${id}`, data)
}

/** 删除摄像头 */
export function deleteCamera(id: number): Promise<{ message: string }> {
  return request.delete(`/v1/video/cameras/${id}`)
}

/** 按区域查询摄像头 */
export function getCamerasByArea(areaCode: string): Promise<CameraItem[]> {
  return request.get(`/v1/video/cameras/by-area/${areaCode}`)
}

/** 按设备查询摄像头 */
export function getCamerasByDevice(deviceId: number): Promise<CameraItem[]> {
  return request.get(`/v1/video/cameras/by-device/${deviceId}`)
}

/** 按告警查询关联摄像头 */
export function getCamerasByAlarm(alarmId: number): Promise<CameraItem[]> {
  return request.get(`/v1/video/cameras/by-alarm/${alarmId}`)
}

// ==================== PTZ / 录像 / 事件 API (Story 10-3) ====================

/** PTZ 控制参数 */
export interface PTZControlParams {
  camera_id: number
  action: string // up/down/left/right/zoom_in/zoom_out/stop
  speed?: number
}

/** 预置位调用参数 */
export interface PresetCallParams {
  camera_id: number
  preset_index: number
}

/** 录像控制参数 */
export interface RecordingParams {
  camera_id: number
  alarm_id?: number
  linkage_execution_id?: number
}

/** 视频事件 */
export interface VideoEventItem {
  id: number
  camera_id: number
  camera_name: string | null
  event_type: string
  trigger_source: string
  alarm_id: number | null
  linkage_execution_id: number | null
  detail: string | null
  operator: string | null
  created_at: string | null
}

/** 云台控制 */
export function ptzControl(data: PTZControlParams): Promise<VideoEventItem> {
  return request.post('/v1/video/ptz/control', data)
}

/** 调用预置位 */
export function callPreset(data: PresetCallParams): Promise<VideoEventItem> {
  return request.post('/v1/video/ptz/preset', data)
}

/** 开始录像 */
export function startRecording(data: RecordingParams): Promise<VideoEventItem> {
  return request.post('/v1/video/recording/start', data)
}

/** 停止录像 */
export function stopRecording(data: RecordingParams): Promise<VideoEventItem> {
  return request.post('/v1/video/recording/stop', data)
}

/** 视频事件列表 */
export function getVideoEvents(params?: PageParams & {
  camera_id?: number
  event_type?: string
}): Promise<PageResponse<VideoEventItem>> {
  return request.get('/v1/video/events', { params })
}

// ==================== 回放 API (Story 10-4) ====================

/** 告警摘要 */
export interface AlarmBrief {
  id: number
  alarm_level: string
  alarm_message: string
  alarm_time: string | null
}

/** 摄像头摘要（回放用） */
export interface CameraBrief {
  id: number
  name: string
  code: string
  rtsp_url: string | null
  hls_url: string | null
  location_description: string | null
}

/** 录像片段 */
export interface RecordingSegment {
  id: number
  camera_id: number
  camera_name: string | null
  start_time: string | null
  end_time: string | null
  alarm_id: number | null
  duration_seconds: number | null
}

/** 告警回放信息 */
export interface PlaybackInfo {
  alarm_info: AlarmBrief
  cameras: CameraBrief[]
  recording_events: VideoEventItem[]
  playback_url_template: string
}

/** 获取告警回放信息 */
export function getPlaybackInfo(alarmId: number): Promise<PlaybackInfo> {
  return request.get(`/v1/video/playback/alarm/${alarmId}`)
}

/** 查询录像片段列表 */
export function getRecordingSegments(params: PageParams & {
  camera_id: number
  start_time?: string
  end_time?: string
}): Promise<PageResponse<RecordingSegment>> {
  return request.get('/v1/video/playback/segments', { params })
}

File: backend/app/schemas/video.py
"""
视频监控 Schema
Story 10-1: 摄像头元数据管理
"""
from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, ConfigDict


# ========== NVR ==========

class NVRCreate(BaseModel):
    """创建 NVR"""
    name: str
    ip_address: str
    port: int = 554
    username: Optional[str] = None
    password: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    max_channels: Optional[int] = None
    description: Optional[str] = None


class NVRUpdate(BaseModel):
    """更新 NVR"""
    name: Optional[str] = None
    ip_address: Optional[str] = None
    port: Optional[int] = None
    username: Optional[str] = None
    password: Optional[str] = None
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    max_channels: Optional[int] = None
    status: Optional[str] = None
    description: Optional[str] = None


class NVRResponse(BaseModel):
    """NVR 响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    ip_address: str
    port: int
    username: Optional[str] = None
    password_masked: Optional[str] = None  # 掩码后的密码
    manufacturer: Optional[str] = None
    model: Optional[str] = None
    max_channels: Optional[int] = None
    status: str
    description: Optional[str] = None
    camera_count: int = 0  # 关联摄像头数量
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ========== CameraPreset ==========

class CameraPresetCreate(BaseModel):
    """创建预置位"""
    preset_index: int
    name: str
    description: Optional[str] = None


class CameraPresetResponse(BaseModel):
    """预置位响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    camera_id: int
    preset_index: int
    name: str
    description: Optional[str] = None


# ========== Camera ==========

class CameraCreate(BaseModel):
    """创建摄像头"""
    name: str
    code: str
    rtsp_url: Optional[str] = None
    onvif_url: Optional[str] = None
    hls_url: Optional[str] = None
    nvr_id: Optional[int] = None
    channel_no: Optional[int] = None
    area_code: Optional[str] = None
    cabinet_id: Optional[int] = None
    device_id: Optional[int] = None
    location_description: Optional[str] = None
    camera_type: str = "dome"
    presets: Optional[List[CameraPresetCreate]] = None


class CameraUpdate(BaseModel):
    """更新摄像头"""
    name: Optional[str] = None
    code: Optional[str] = None
    rtsp_url: Optional[str] = None
    onvif_url: Optional[str] = None
    hls_url: Optional[str] = None
    nvr_id: Optional[int] = None
    channel_no: Optional[int] = None
    area_code: Optional[str] = None
    cabinet_id: Optional[int] = None
    device_id: Optional[int] = None
    location_description: Optional[str] = None
    camera_type: Optional[str] = None
    status: Optional[str] = None
    is_enabled: Optional[bool] = None
    presets: Optional[List[CameraPresetCreate]] = None


class CameraResponse(BaseModel):
    """摄像头响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    code: str
    rtsp_url: Optional[str] = None
    onvif_url: Optional[str] = None
    hls_url: Optional[str] = None
    nvr_id: Optional[int] = None
    nvr_name: Optional[str] = None  # 关联 NVR 名称
    channel_no: Optional[int] = None
    area_code: Optional[str] = None
    cabinet_id: Optional[int] = None
    device_id: Optional[int] = None
    location_description: Optional[str] = None
    camera_type: str
    status: str
    is_enabled: bool
    presets: List[CameraPresetResponse] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


# ========== VideoEvent ==========

class PTZControlRequest(BaseModel):
    """云台控制请求"""
    camera_id: int
    action: str  # up/down/left/right/zoom_in/zoom_out/stop
    speed: int = 5  # 1-10


class PresetCallRequest(BaseModel):
    """预置位调用请求"""
    camera_id: int
    preset_index: int


class RecordingRequest(BaseModel):
    """录像控制请求"""
    camera_id: int
    alarm_id: Optional[int] = None
    linkage_execution_id: Optional[int] = None


class VideoEventResponse(BaseModel):
    """视频事件响应"""
    model_config = ConfigDict(from_attributes=True)

    id: int
    camera_id: int
    camera_name: Optional[str] = None
    event_type: str
    trigger_source: str
    alarm_id: Optional[int] = None
    linkage_execution_id: Optional[int] = None
    detail: Optional[str] = None
    operator: Optional[str] = None
    created_at: Optional[datetime] = None


# ========== Playback (Story 10-4) ==========

class AlarmBrief(BaseModel):
    """告警摘要"""
    id: int
    alarm_level: str
    alarm_message: str
    alarm_time: Optional[datetime] = None


class CameraBrief(BaseModel):
    """摄像头摘要（回放用）"""
    id: int
    name: str
    code: str
    rtsp_url: Optional[str] = None
    hls_url: Optional[str] = None
    location_description: Optional[str] = None


class RecordingSegmentResponse(BaseModel):
    """录像片段"""
    id: int
    camera_id: int
    camera_name: Optional[str] = None
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    alarm_id: Optional[int] = None
    duration_seconds: Optional[int] = None


class PlaybackInfoResponse(BaseModel):
    """告警回放信息"""
    alarm_info: AlarmBrief
    cameras: List[CameraBrief]
    recording_events: List[VideoEventResponse]
    playback_url_template: str
